backend/module/chaos/serverexperimentation/xds/xds_cache_test.go

```diff
@@ -279,3 +291,131 @@ func TestComputeVersionReturnValue(t *testing.T) {
 	checksum2, _ := computeChecksum(mockRuntime2)
 	assert.NotEqual(t, checksum, checksum2)
 }
+
+func TestSetSnapshotECDSInternalFault(t *testing.T) {
+	testCache := gcpCacheV3.NewSnapshotCache(false, gcpCacheV3.IDHash{}, nil)
+	testCluster := "serviceA"
+	var downstreamCluster string
+	mockExperimentList := mockGenerateFaultData(t)
+
+	var testClusterFaults []*experimentation.Experiment
+	for _, experiment := range mockExperimentList {
+		config := &serverexperimentation.HTTPFaultConfig{}
+		err := ptypes.UnmarshalAny(experiment.GetConfig(), config)
+		if err != nil {
+			t.Errorf("unmarshalAny failed %v", err)
+		}
+
+		upstream, downstream, err := getClusterPair(config)
+		assert.NoError(t, err)
+
+		downstreamCluster = downstream
+
+		// Test Internal Fault
+		if upstream == testCluster {
+			testClusterFaults = append(testClusterFaults, experiment)
+
+			// Since ECDS can only perform one experiment per cluster
+			break
+		}
+	}
+
+	resources := make(map[gcpTypes.ResponseType][]gcpTypes.ResourceWithTtl)
+	resources[gcpTypes.ExtensionConfig] = generateECDSResource(testClusterFaults, nil, zap.NewNop().Sugar())
+	assert.Nil(t, resources[gcpTypes.ExtensionConfig][0].Ttl)
+	err := setSnapshot(resources, testCluster, testCache, zap.NewNop().Sugar())
+	if err != nil {
+		t.Errorf("setSnapshot failed %v", err)
+	}
+
+	snapshot, err := testCache.GetSnapshot(testCluster)
+	if err != nil {
+		t.Errorf("Snapshot not found for cluster %s", testCluster)
+	}
+
+	ecdsResources := snapshot.Resources[gcpTypes.ExtensionConfig].Items
+	if ecdsResources == nil {
+		t.Errorf("no resources")
+	}
+
+	assert.Contains(t, ecdsResources, FaultFilterConfigNameForInternalFault)
+
+	extensionConfig := ecdsResources[FaultFilterConfigNameForInternalFault].Resource.(*gcpCoreV3.TypedExtensionConfig)
+
+	httpFaultFilter := &gcpFilterFault.HTTPFault{}
+	err = ptypes.UnmarshalAny(extensionConfig.TypedConfig, httpFaultFilter)
+	if err != nil {
+		t.Errorf("Unable to unmarshall typed config")
+	}
+
+	assert.EqualValues(t, 10, httpFaultFilter.GetAbort().GetPercentage().GetNumerator())
+	assert.EqualValues(t, 404, httpFaultFilter.GetAbort().GetHttpStatus())
+	assert.EqualValues(t, 0, httpFaultFilter.GetDelay().GetPercentage().GetNumerator())
+	assert.EqualValues(t, 0, httpFaultFilter.GetDelay().GetFixedDelay().GetSeconds())
+	assert.EqualValues(t, 1, len(httpFaultFilter.GetHeaders()))
+	assert.EqualValues(t, EnvoyDownstreamServiceClusterHeader, httpFaultFilter.GetHeaders()[0].GetName())
+	assert.EqualValues(t, downstreamCluster, httpFaultFilter.GetHeaders()[0].GetExactMatch())
```

Comment:
Might find that proto equality of a fully constructed proto ends up being easier to read + ensures that we're not setting fields that we don't expect to set